src/concrete/ml/quantization/quantized_ops.py

```diff
@@ -217,20 +222,113 @@ def q_impl(
         # Here we follow Eq.7 in https://arxiv.org/abs/1712.05877 to split the core computation
         # from the zero points and scales.
 
-        p = weights_q_values.shape[0]
+        p = input2_q_values.shape[-2]
+
+        # FIXME: https://github.com/zama-ai/concrete-internal/issues/512
+        def enc_mul(x, y):
+            """Encrypted multiplication.
+
+            Args:
+                x (numpy.ndarray): first input numpy array
+                y (numpy.ndarray): second input numpy array
+
+            Returns:
+                numpy.ndarray: result of the multiplication
+            """
+            with tag("pbs_multiplication"):
+                add = x + y
+                sub = x - y
+                with tag(self.op_instance_name + ".pbs_matmul_rounding_add"):
+                    # Apply Concrete rounding (if relevant)
+                    add = self.cnp_round(add, calibrate_rounding, key="add")
+                with tag(self.op_instance_name + ".pbs_matmul_rounding_sub"):
+                    # Apply Concrete rounding (if relevant)
+                    sub = self.cnp_round(sub, calibrate_rounding, key="sub")
+
+                add_pow = (add.astype(numpy.float64)) ** 2
+                sub_pow = (sub.astype(numpy.float64)) ** 2
+                add_pow_divide = (add_pow / 4.0).astype(numpy.int64)
+                sub_pow_divide = (sub_pow / 4.0).astype(numpy.int64)
+
+            return add_pow_divide - sub_pow_divide
+
+        # FIXME: https://github.com/zama-ai/concrete-internal/issues/512
+        def matmul(a, b):
+            with tag("Encrypted MatMul"):
```

Comment:
tags with space.. ugh. please please use lowercase and `_`

Reply:
Done